checkbox-support/checkbox_support/helpers/retry.py

```diff
@@ -0,0 +1,76 @@
+# This file is part of Checkbox.
+#
+# Copyright 2024 Canonical Ltd.
+# Written by:
+#   Pierre Equoy <pierre.equoy@canonical.com>
+#
+# Checkbox is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3,
+# as published by the Free Software Foundation.
+#
+# Checkbox is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Checkbox.  If not, see <http://www.gnu.org/licenses/>.
+"""
+checkbox_support.helpers.retry
+=============================================
+
+Utility class providing functionalities to let functions retry with a
+delay, backoff and jitter.
+"""
+import functools
+import random
+import time
+
+
+def retry(
+    f=None, *, max_attempts=5, initial_delay=1, backoff_factor=2, max_delay=60
```

Comment:
If that `f=None` is to allow this style of calling: `retry(max_attempts=3)(abc)((1, ))`. Dont. It doesn't make much sense to do this this way. I prefer to just have a helper function that does this (like we did for timeout)

Reply:
else please explain it, I dont get it

Reply:
This syntax is required if you want to pass optional arguments to a decorator, as explained in this [Real Python article.](https://realpython.com/primer-on-python-decorators/#creating-decorators-with-optional-arguments) Do you have a better idea on how to achieve this?

Reply:
In the end I decided optional arguments were a bad idea, so I went with positional arguments that have to be defined whenever you call the `@retry` decorator.